src/blockencodings.cpp

```diff
@@ -125,20 +128,32 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
                     txn_available[idit->second] = vTxHashes[i].second->GetSharedTx();
                     have_txn[idit->second]  = true;
                     mempool_count++;
+                    missing_txs--;
                 } else {
                     // If we find two mempool txn that match the short id, just request it.
                     // This should be rare enough that the extra bandwidth doesn't matter,
                     // but eating a round-trip due to FillBlock failure would be annoying
                     if (txn_available[idit->second]) {
                         txn_available[idit->second].reset();
-                        mempool_count--;
+                        shortid_collision = true;
                     }
                 }
             }
-            // Though ideally we'd continue scanning for the two-txn-match-shortid case,
-            // the performance win of an early exit here is too good to pass up and worth
-            // the extra risk.
-            if (mempool_count == shorttxids.size()) {
+            if (missing_txs == 0 && !shortid_collision) {
+                // We've found transactions matching all of the shortids and haven't yet
+                // found a shortid collision. We could continue scanning through the
+                // mempool and extra pool in case there's a shortid collision later,
+                // but running GetShortID() on the entire mempool and extra pool is an
+                // expensive operation.
```

Comment:
Wouldn't call it "expensive" per se, but weighing a 1 in a million case where we use more bandwidth against something that takes 10ms...I'll take the 1 in a million.